lib/jwe.js

```diff
@@ -0,0 +1,511 @@
+      var decipher = crypto.createDecipheriv(mapper[header.enc], cek, iv);
+      decipher.setAAD(aad);
+      decipher.setAuthTag(authTag);
+      content = decipher.update(cipherText);
+      content = Buffer.concat([content, decipher.final()]);
+    } else {
+      content = decrypt_AES_CBC_HMAC_SHA2(mapper[header.enc], cek, cipherText, iv, aad, authTag);    
+    }
+
+    return { 'error': null, 'content': content.toString() };
+  } catch (ex) {
+    return { 'error': ex, 'content': null };
+  }
+};
+
+/**
+ * decrypt content function for the case where alg is dir
+ *
+ * @param{Object}  header: header
+ * @param{Buffer}  jweKeyStore: list of keys
+ * @param{Buffer}  cipherText: encrypted content
+ * @param{Buffer}  iv: initialization vector
+ * @param{Buffer}  authTag: authentication tag
+ * @param{Buffer}  aad: additional authentication information
+ * @return{Function}  log: logging function
+ */
+var decrytContentForDir = (header, jweKeyStore, cipherText, iv, authTag, aad, log) => {
+  var key = null;
+
+  // try the key with the corresponding kid
+  if (header.kid) {
+    for (var i = 0; i < jweKeyStore.length; i++) {
+      if (header.kid === jweKeyStore[i].kid && jweKeyStore[i].kty === 'oct') {
+        key = jweKeyStore[i];
+        log.info('found a key matching kid: ' + header.kid);
+        break;
+      }
+    }
+
+    if (key)
+      return decrytContent(header, base64url.toBuffer(key.k), cipherText, iv, authTag, aad, log);
+  }
+
+  log.info('In decrytContentForDir: we cannout find a key using kid, now we try all the possible keys');
+
+  // kid matching failed, now we try every possible key
+  for(var i = 0; i < jweKeyStore.length; i++) {
+    if (jweKeyStore[i].kty === 'oct') {
+      var result = decrytContent(header, base64url.toBuffer(jweKeyStore[i].k), cipherText, iv, authTag, aad, log);
+      if (!result.error)
+        return result;
+    }
+  }
+
+  log.info('In decrytContentForDir: tried all keys to decrypt the content but all failed');
+  return { error: new Error('In decrytContentForDir: tried all keys to decrypt the content but all failed'), content_result: null };
+};
+
+/**
+ * The main decryption function
+ *
+ * @param{String}  jweString:  JWE id_token
+ * @param{Object}  jweKeyStore: list of keys
+ * @return{Function}  log: logging function
+ * @return{Function}  callback: callback function
+ */
+exports.decryt = (jweString, jweKeyStore, log, callback) => {
+  /****************************************************************************
+   *   JWE compact format structure
+   ****************************************************************************
+   * BASE64URL(UTF8(JWE Protected Header)) || '.' ||
+   * BASE64URL(JWE Encrypted Key) || '.' || 
+   * BASE64URL(JWE Initialization Vector) || '.' || 
+   * BASE64URL(JWE Ciphertext) || '.' || 
+   * BASE64URL(JWE Authentication Tag)
+   ***************************************************************************/
+  var parts = jweString.split('.');
```

Comment:
the number of parts is already validated in oidcstrategy _decryptJWE function. Now I added the number of parts validation here too. Now also added error checking for header parsing (since if it is not a json passport will throw), and a couple of test cases where some parts are empty string.